### PR TITLE
Add optional args in chip-tool and end point optional arg

### PR DESCRIPTION
**Pull request by krypton36**
#### Problem
- Add optional arguments to chip-tool
- Add ability to set EndPoint ID on chip tool command run.
- Issued [Here #11100 ](https://github.com/project-chip/connectedhomeip/issues/11102)

#### Change overview
In chip-tool add an argument that will allow adding optional arguments. Optional arguments. ZAP file updates to add the optional flag to false.

#### Testing
- Tested Zap generation with the generate command. 
- Testsed compilation with ninja
- Tested chip-tool optional arguments.
- Tested pairing command.
- Tested tests command in chip-tool.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking given this is a test tool.

File: examples/chip-tool/commands/common/Command.cpp [hunk 1]
```diff
@@ -32,6 +32,10 @@
 #include <lib/support/ScopedBuffer.h>
 #include <lib/support/logging/CHIPLogging.h>
 
+#define NO_OPTIONAL_ARGS 0
+#define OPTIONAL_ARGS_OFFSET 2
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:36]: Please document that this is because each optional arg is two entries in argv: name and value.

File: examples/chip-tool/commands/common/Command.cpp [hunk 2]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+    int argsCount       = static_cast<int>(mArgs.size());
+    int optionalArgC    = argc - argsCount;
+    int expectedArgs    = optionalArgC;
+    int optInits        = 0;
+
+    // If no optional args passed in, in return true.
+    if (optionalArgC == NO_OPTIONAL_ARGS)
+        return true;
+
+    // If too many optinoal args passed in, return false.
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:74]:
```suggestion
    // If too many optional args passed in, return false.
```

File: examples/chip-tool/commands/common/Command.cpp [hunk 3]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+    if (optArgsCount * OPTIONAL_ARGS_OFFSET < optionalArgC)
+    {
+        ChipLogError(chipTool, "Too Many arguments passed in.");
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:77]:
```suggestion
        ChipLogError(chipTool, "Too many arguments passed in.");
```

File: examples/chip-tool/commands/common/Command.h [hunk 4]
```diff
@@ -91,10 +93,13 @@ class Command
     const char * GetName(void) const { return mName; }
     const char * GetAttribute(void) const;
     const char * GetArgumentName(size_t index) const;
+    const char * GetOptionalArgumentName(size_t index) const;
     size_t GetArgumentsCount(void) const { return mArgs.size(); }
+    size_t GetOptionalArgumentsCount(void) const { return mOptionalArgs.size(); }
 
     bool InitArguments(int argc, char ** argv);
-    size_t AddArgument(const char * name, const char * value);
+    bool InitOptionalArguments(int & argc, char ** argv);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.h:101]: Need to document the semantics, especially of the return value and the in/out parameter.

File: examples/chip-tool/commands/common/Command.cpp [hunk 5]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
     return isValidCommand;
 }
 
+bool Command::InitOptionalArguments(int & argc, char ** argv)
+{
+    bool isValidCommand = false;
+    int optArgsCount    = static_cast<int>(mOptionalArgs.size());
+    int argsCount       = static_cast<int>(mArgs.size());
+    int optionalArgC    = argc - argsCount;
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:66]: So this can end up negative, right?

File: examples/chip-tool/commands/common/Command.cpp [hunk 6]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+        return true;
+
+    // If too many optinoal args passed in, return false.
+    if (optArgsCount * OPTIONAL_ARGS_OFFSET < optionalArgC)
+    {
+        ChipLogError(chipTool, "Too Many arguments passed in.");
+        return false;
+    }
+    // Iterate the total configured optional args
+    for (int i = 0; i < optArgsCount; i++)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:81]: So ideally optArgsCount and `i` would both be `size_t` here.  Any reason not to give them those types?

File: examples/chip-tool/commands/common/Command.cpp [hunk 7]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+    {
+        Argument arg = mOptionalArgs.at((size_t) i);
+        // Check to see if the argument matches what has been configured.
+        if (argv[optInits * OPTIONAL_ARGS_OFFSET + argsCount] != nullptr &&
+            strcmp(arg.name, argv[optInits * OPTIONAL_ARGS_OFFSET + argsCount]) == 0)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:86]: What's keeping this from indexing out of bounds into argv?  I don't see anything preventing that.  Most simply, if `optionalArgC` == 1 but we have two possible optional args, once we find the first one the second one will index out of bounds.

The `optionalArgC == NO_OPTIONAL_ARGS` works around this for the special case of no optional args passed in at all, which is why this might not have been caught....  We should probably take out that special-case and have better testing for "we have exhausted all the args", and that testing needs to be inside this loop.

File: examples/chip-tool/commands/common/Command.cpp [hunk 8]
```diff
@@ -32,6 +32,10 @@
 #include <lib/support/ScopedBuffer.h>
 #include <lib/support/logging/CHIPLogging.h>
 
+#define NO_OPTIONAL_ARGS 0
+#define OPTIONAL_ARGS_OFFSET 2
+#define OPTIONAL_VALUE_OFFSET 1
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:37]: This is because the value is right after the name?

I think it might be clearer to just have functions for getting the Nth optional arg name and the Nth optional arg value (which take argv, the number of non-optional args expected, and N).

File: examples/chip-tool/commands/common/Command.cpp [hunk 9]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+    if (optionalArgC == NO_OPTIONAL_ARGS)
+        return true;
+
+    // If too many optinoal args passed in, return false.
+    if (optArgsCount * OPTIONAL_ARGS_OFFSET < optionalArgC)
+    {
+        ChipLogError(chipTool, "Too Many arguments passed in.");
+        return false;
+    }
+    // Iterate the total configured optional args
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:80]: So this setup requires the optional args to be provided in the same order as our storage, right?  That seems fine, but we might want to document that.

File: examples/chip-tool/commands/common/Command.cpp [hunk 10]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+                ExitNow();
+            }
+            optInits++; // Increment the initialized optional args
+        }
+    }
+
+    ChipLogError(chipTool, "Exit Loop");
+    // Subract the total initialized optional args passed in and reset pointer.
+    argc -= optInits * OPTIONAL_ARGS_OFFSET;
+    argv[argsCount] = nullptr;
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:101]: Why is this writing into `argv[argsCount]`?

File: examples/chip-tool/commands/common/Command.cpp [hunk 11]
```diff
@@ -54,6 +58,59 @@ bool Command::InitArguments(int argc, char ** argv)
...
+    // If the expected optional args are not inittialized, return false. ie wrong names.
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Command.cpp:102]:
```suggestion
    // If the expected optional args are not initialized, return false. ie wrong names.
```

File: examples/chip-tool/commands/common/Commands.cpp [hunk 12]
```diff
@@ -107,8 +107,14 @@ CHIP_ERROR Commands::RunCommand(int argc, char ** argv)
             return CHIP_ERROR_INVALID_ARGUMENT;
         }
     }
+    argc -= 3;
+    if (!command->InitOptionalArguments(argc, &argv[3]))
+    {
+        ShowCommand(argv[0], argv[1], command);
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/common/Commands.cpp:115]: So here's a question.  Does it make sense to init the required arguments first, then on success ask the command how many non-optional args it ended up with, decrement argc by that, offset by that into argv and call `InitOptionalArguments`?  Then `InitOptionalArguments` won't need an in/out param and the indexing will be much simpler...

File: examples/chip-tool/commands/tests/TestCommand.h [hunk 13]
```diff
@@ -27,15 +27,19 @@
 #include <type_traits>
 #include <zap-generated/tests/CHIPClustersTest.h>
 
+// Limits on endpoint values.
+#define CHIP_ZCL_ENDPOINT_MIN 0x00
+#define CHIP_ZCL_ENDPOINT_MAX 0xF0
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/tests/TestCommand.h:32]: Why are these here?

File: examples/chip-tool/templates/partials/test_cluster.zapt [hunk 14]
```diff
@@ -115,7 +115,13 @@ class {{filename}}: public TestCommand
     {{else if isWait}}
     CHIP_ERROR {{>testCommand}}()
     {
-      ChipLogError(chipTool, "[Endpoint: {{endpoint}} Cluster: {{cluster}} {{#if isAttribute}}Attribute: {{attribute}}{{else}}Command: {{wait}}{{/if}}] {{label}}");
+      uint8_t endPoint = {{endpoint}};
+      Argument arg     = mOptionalArgs.at((size_t) 0);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/templates/partials/test_cluster.zapt:119]:
```suggestion
      Argument arg     = mOptionalArgs.at(0);
```

File: examples/chip-tool/templates/partials/test_cluster.zapt [hunk 15]
```diff
@@ -131,7 +137,14 @@ class {{filename}}: public TestCommand
...
     {
         chip::Controller::{{asUpperCamelCase cluster}}ClusterTest cluster;
-        cluster.Associate(mDevice, {{endpoint}});
+        uint8_t endPoint = {{endpoint}};
+        Argument arg     = mOptionalArgs.at((size_t) 0);
+        if (arg.isProvided)
+        {
+          endPoint = mEndPointId;
+        }
+        ChipLogProgress(chipTool, "Was Provided: %s and Enpoint Value: %d", arg.isProvided ? "true" : "false", endPoint);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/templates/partials/test_cluster.zapt:146]:
```suggestion
        ChipLogProgress(chipTool, "Was Provided: %s and Endpoint Value: %d", arg.isProvided ? "true" : "false", endPoint);
```

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by bzbarsky-apple** [comment]
@krypton36 it looks like @vivien-apple added something along these lines (but with different semantics?  Not sure whether it works right?) in #11729, so at the very least this will need to get rebased on top of that.... :(

**Comment by krypton36** [comment]
I am closing this PR and enhancing Vivian's PR for the optional arguments to avoid conflict nightmares.

**Comment by bzbarsky-apple** [comment]
Makes sense, but I wish we were talking to each other more....Ah, well.
